department-of-veterans-affairs/vets-api
# [54884] Sign in Service dual-certificate support

## Summary

This PR makes two changes to expand the number of `vets-api` / Sign in Service certificates that support two simultaneous certificates.
* Sign in Service will check for the presence of `Settings.sign_in.jwt_old_encode_key` when populating the response to the public certificates display route; if this setting is populated the route will respond with an array of both the current public key and the previous one (tied to the `jwt_old_encode_key` setting).
* SiS will also check for live public keys from ID.me when decoding the `user_info` request made during authentication. The first time the ID.me public cert is needed in an environment the `Idme::Configuration` library will call out to the ID.me public certs endpoint set at `Settings.idme.oauth_public_key_url`. Future requests will pull the keys from the `IdmePublicKey` class where they are stored for the remainder of the `vets-api` instance lifespan.

## Related issue(s)
- department-of-veterans-affairs/va.gov-team#54884
- DevOps PR: https://github.com/department-of-veterans-affairs/devops/pull/13293
- VSP Infra PR: https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/2157
- Production VSP Infra PR: https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/2165
- the following values have been added to AWS PS:
  - `/dsva-vagov/vets-api/dev/idme/oauth_public_key_url` => https://api.idmelabs.com/oidc/.well-known/jwks
  - `/dsva-vagov/vets-api/staging/idme_oauth_public_key_ur` => https://api.idmelabs.com/oidc/.well-known/jwks
  - `/dsva-vagov/vets-api/prod/idme/oauth_public_key_url` => https://api.id.me/oidc/.well-known/jwks

## Testing done

- Relevant unit specs added & updated.

### Testing Instructions - ID.me public keys

- Place a breakpoint in the `jwt_decode_public_key` function in `lib/h/idme/configuration.rb` (after the fetcher performs), then perform a SiS ID.me authentication.
- At your breakpoint, verify that `@idme_public_keys` has been set to an RS256 key.

### Testing Instructions - SiS public keys

- Navigate to `http://localhost:3000/sign_in/openid_connect/certs` - you should see two separate RS256 keys.
- Set your local `Settings.sign_in.jwt_old_encode_key` to `false`, then restart your server & refresh the route. You should now see only one RS256 key displayed.


## What areas of the site does it impact?
- ID.me authentication
- SiS client authentication

## Acceptance criteria


## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_
